File: mini-nn-go/nn/optimizer.go
package nn

import "math"

// OptimizerType represents the type of optimizer.
type OptimizerType int

const (
	SGDType OptimizerType = iota
	SGDMomentumType
	AdamType
)

// Optimizer interface for all optimizers.
type Optimizer interface {
	Step(layers []*Dense)
	LearningRate() float64
}

// SGD implements stochastic gradient descent.
type SGD struct {
	LR float64
}

// NewSGD creates an SGD optimizer.
func NewSGD(lr float64) *SGD {
	return &SGD{LR: lr}
}

// Step updates weights using gradients.
func (o *SGD) Step(layers []*Dense) {
	for _, layer := range layers {
		// Update weights: W = W - lr * grad_W
		r, c := layer.Weights.Shape()
		for i := 0; i < r; i++ {
			for j := 0; j < c; j++ {
				w := layer.Weights.At(i, j)
				g := layer.GradWeights.At(i, j)
				layer.Weights.Set(i, j, w-o.LR*g)
			}
		}

		// Update biases: b = b - lr * grad_b
		for i := range layer.Bias {
			layer.Bias[i] -= o.LR * layer.GradBias[i]
		}
	}
}

// LearningRate returns the learning rate.
func (o *SGD) LearningRate() float64 {
	return o.LR
}

// SGDMomentum implements SGD with momentum.
type SGDMomentum struct {
	LR       float64
	Momentum float64
	vWeights []*Tensor
	vBias    [][]float64
	init     bool
}

// NewSGDMomentum creates an SGD optimizer with momentum.
func NewSGDMomentum(lr, momentum float64) *SGDMomentum {
	return &SGDMomentum{LR: lr, Momentum: momentum}
}

// Step updates weights using gradients with momentum.
func (o *SGDMomentum) Step(layers []*Dense) {
	if !o.init {
		o.vWeights = make([]*Tensor, len(layers))
		o.vBias = make([][]float64, len(layers))
		for i, layer := range layers {
			o.vWeights[i] = Zeros(layer.InFeatures, layer.OutFeatures)
			o.vBias[i] = make([]float64, layer.OutFeatures)
		}
		o.init = true
	}

	for i, layer := range layers {
		r, c := layer.Weights.Shape()

		// Update velocity and weights
		for ri := 0; ri < r; ri++ {
			for ci := 0; ci < c; ci++ {
				v := o.Momentum*o.vWeights[i].At(ri, ci) - o.LR*layer.GradWeights.At(ri, ci)
				o.vWeights[i].Set(ri, ci, v)
				layer.Weights.Set(ri, ci, layer.Weights.At(ri, ci)+v)
			}
		}

		// Update bias velocity and biases
		for bi := range layer.Bias {
			o.vBias[i][bi] = o.Momentum*o.vBias[i][bi] - o.LR*layer.GradBias[bi]
			layer.Bias[bi] += o.vBias[i][bi]
		}
	}
}

// LearningRate returns the learning rate.
func (o *SGDMomentum) LearningRate() float64 {
	return o.LR
}

// Adam implements the Adam optimizer.
type Adam struct {
	LR      float64
	Beta1   float64
	Beta2   float64
	Epsilon float64
	mW      []*Tensor // First moment for weights
	vW      []*Tensor // Second moment for weights
	mB      [][]float64
	vB      [][]float64
	t       int // Time step
	init    bool
}

// NewAdam creates an Adam optimizer with default hyperparameters.
func NewAdam(lr float64) *Adam {
	return &Adam{
		LR:      lr,
		Beta1:   0.9,
		Beta2:   0.999,
		Epsilon: 1e-8,
	}
}

// NewAdamFull creates an Adam optimizer with custom hyperparameters.
func NewAdamFull(lr, beta1, beta2, epsilon float64) *Adam {
	return &Adam{
		LR:      lr,
		Beta1:   beta1,
		Beta2:   beta2,
		Epsilon: epsilon,
	}
}

// Step updates weights using Adam algorithm.
func (o *Adam) Step(layers []*Dense) {
	if !o.init {
		o.mW = make([]*Tensor, len(layers))
		o.vW = make([]*Tensor, len(layers))
		o.mB = make([][]float64, len(layers))
		o.vB = make([][]float64, len(layers))
		for i, layer := range layers {
			o.mW[i] = Zeros(layer.InFeatures, layer.OutFeatures)
			o.vW[i] = Zeros(layer.InFeatures, layer.OutFeatures)
			o.mB[i] = make([]float64, layer.OutFeatures)
			o.vB[i] = make([]float64, layer.OutFeatures)
		}
		o.init = true
	}

	o.t++
	biasCorr1 := 1.0 - math.Pow(o.Beta1, float64(o.t))
	biasCorr2 := 1.0 - math.Pow(o.Beta2, float64(o.t))

	for i, layer := range layers {
		r, c := layer.Weights.Shape()

		// Update weights
		for ri := 0; ri < r; ri++ {
			for ci := 0; ci < c; ci++ {
				g := layer.GradWeights.At(ri, ci)

				// Update biased first moment estimate
				m := o.Beta1*o.mW[i].At(ri, ci) + (1-o.Beta1)*g
				o.mW[i].Set(ri, ci, m)

				// Update biased second raw moment estimate
				v := o.Beta2*o.vW[i].At(ri, ci) + (1-o.Beta2)*g*g
				o.vW[i].Set(ri, ci, v)

				// Bias-corrected estimates
				mHat := m / biasCorr1
				vHat := v / biasCorr2

				// Update weights
				w := layer.Weights.At(ri, ci)
				layer.Weights.Set(ri, ci, w-o.LR*mHat/(math.Sqrt(vHat)+o.Epsilon))
			}
		}

		// Update biases
		for bi := range layer.Bias {
			g := layer.GradBias[bi]

			o.mB[i][bi] = o.Beta1*o.mB[i][bi] + (1-o.Beta1)*g
			o.vB[i][bi] = o.Beta2*o.vB[i][bi] + (1-o.Beta2)*g*g

			mHat := o.mB[i][bi] / biasCorr1
			vHat := o.vB[i][bi] / biasCorr2

			layer.Bias[bi] -= o.LR * mHat / (math.Sqrt(vHat) + o.Epsilon)
		}
	}
}

// LearningRate returns the learning rate.
func (o *Adam) LearningRate() float64 {
	return o.LR
}

File: mini-nn-python/mini_nn/optimizer.py
"""Optimizers for neural network training."""

import numpy as np
from abc import ABC, abstractmethod
from typing import List, Dict
from .tensor import Tensor
from .layer import DenseLayer


class Optimizer(ABC):
    """Base class for optimizers."""

    @abstractmethod
    def step(self, layers: List[DenseLayer]):
        """Update layer parameters based on gradients."""
        pass


class SGDOptimizer(Optimizer):
    """Stochastic Gradient Descent optimizer."""

    def __init__(self, learning_rate: float = 0.01):
        """Initialize SGD optimizer."""
        self.learning_rate = learning_rate

    def step(self, layers: List[DenseLayer]):
        """Update parameters using SGD."""
        for layer in layers:
            weights, bias = layer.get_parameters()
            weight_grad, bias_grad = layer.get_gradients()
            
            # W = W - lr * dW
            new_weights = weights.sub(weight_grad.scale(self.learning_rate))
            # b = b - lr * db
            new_bias = bias.sub(bias_grad.scale(self.learning_rate))
            
            layer.set_parameters(new_weights, new_bias)


class MomentumOptimizer(Optimizer):
    """SGD with Momentum optimizer."""

    def __init__(self, learning_rate: float = 0.01, momentum: float = 0.9):
        """Initialize Momentum optimizer."""
        self.learning_rate = learning_rate
        self.momentum = momentum
        self.velocity_w: Dict[int, Tensor] = {}
        self.velocity_b: Dict[int, Tensor] = {}

    def step(self, layers: List[DenseLayer]):
        """Update parameters using Momentum."""
        for i, layer in enumerate(layers):
            weights, bias = layer.get_parameters()
            weight_grad, bias_grad = layer.get_gradients()
            
            # Initialize velocity if needed
            if i not in self.velocity_w:
                self.velocity_w[i] = Tensor.zeros(*weights.shape)
                self.velocity_b[i] = Tensor.zeros(*bias.shape)
            
            # Update velocity
            vw = self.velocity_w[i].scale(self.momentum).sub(
                weight_grad.scale(self.learning_rate)
            )
            vb = self.velocity_b[i].scale(self.momentum).sub(
                bias_grad.scale(self.learning_rate)
            )
            
            self.velocity_w[i] = vw
            self.velocity_b[i] = vb
            
            # Update parameters
            new_weights = weights.add(vw)
            new_bias = bias.add(vb)
            
            layer.set_parameters(new_weights, new_bias)


class AdamOptimizer(Optimizer):
    """Adam optimizer."""

    def __init__(
        self,
        learning_rate: float = 0.001,
        beta1: float = 0.9,
        beta2: float = 0.999,
        epsilon: float = 1e-8,
    ):
        """Initialize Adam optimizer."""
        self.learning_rate = learning_rate
        self.beta1 = beta1
        self.beta2 = beta2
        self.epsilon = epsilon
        self.t = 0
        
        # First and second moment estimates
        self.m_w: Dict[int, Tensor] = {}
        self.v_w: Dict[int, Tensor] = {}
        self.m_b: Dict[int, Tensor] = {}
        self.v_b: Dict[int, Tensor] = {}

    def step(self, layers: List[DenseLayer]):
        """Update parameters using Adam."""
        self.t += 1
        
        for i, layer in enumerate(layers):
            weights, bias = layer.get_parameters()
            weight_grad, bias_grad = layer.get_gradients()
            
            # Initialize moments if needed
            if i not in self.m_w:
                self.m_w[i] = Tensor.zeros(*weights.shape)
                self.v_w[i] = Tensor.zeros(*weights.shape)
                self.m_b[i] = Tensor.zeros(*bias.shape)
                self.v_b[i] = Tensor.zeros(*bias.shape)
            
            # Update weights
            new_weights = self._update_param(
                weights, weight_grad, self.m_w[i], self.v_w[i]
            )
            
            # Update bias
            new_bias = self._update_param(
                bias, bias_grad, self.m_b[i], self.v_b[i]
            )
            
            layer.set_parameters(new_weights, new_bias)

    def _update_param(
        self, param: Tensor, grad: Tensor, m: Tensor, v: Tensor
    ) -> Tensor:
        """Update a single parameter using Adam."""
        g = grad.numpy()
        m_data = m.numpy()
        v_data = v.numpy()
        
        # Update biased first moment estimate
        m_data[:] = self.beta1 * m_data + (1 - self.beta1) * g
        
        # Update biased second raw moment estimate
        v_data[:] = self.beta2 * v_data + (1 - self.beta2) * g * g
        
        # Compute bias-corrected estimates
        m_hat = m_data / (1 - self.beta1 ** self.t)
        v_hat = v_data / (1 - self.beta2 ** self.t)
        
        # Update parameter
        update = self.learning_rate * m_hat / (np.sqrt(v_hat) + self.epsilon)
        return Tensor(param.numpy() - update)
